include/nil/crypto3/zk/snark/systems/plonk/kate/proving_key.hpp

```diff
@@ -0,0 +1,299 @@
+                        // However, as explained in
+                        // (./src/aztec/plonk/proof_system/prover/prover.cpp/ProverBase::compute_quotient_pre_commitment),
+                        // for standard plonk (program_width = 3) and number of roots cut out of the vanishing
+                        // polynomial is 4, the degree of the quotient polynomial t(X) is 3n. Thus, the number of
+                        // coefficients in t_{high} is (n + 1). But our prover algorithm assumes that each of t_{low},
+                        // t_{mid}, t_{high} is of degree (n - 1) (i.e. n coefficients in each). Note that:
+                        // deg(W_{\script{z}}) = max{ deg(t_{low}), deg(t_{mid}), deg(t_{high}), deg(a), deg(b), ... }
+                        // => deg(W_{\script{z}}) = n + 1 when program_width is 3!
+                        // Therefore, when program_width is 3, we need to allow the degree of the opening polynomial to
+                        // be (n + 1) and NOT n.
+                        //
+                        opening_poly = math::polynomial(n, n);
+                        shifted_opening_poly = math::polynomial(n, n);
+                        linear_poly = math::polynomial(n, n);
+
+                        quotient_mid = math::polynomial(2 * n, 2 * n);
+                        quotient_large = math::polynomial(4 * n, 4 * n);
+
+                        memset((void *)&opening_poly[0], 0x00, scalar_bytes * n);
+                        memset((void *)&shifted_opening_poly[0], 0x00, scalar_bytes * n);
+                        memset((void *)&linear_poly[0], 0x00, scalar_bytes * n);
+                        memset((void *)&quotient_mid[0], 0x00, scalar_bytes * 2 * n);
+                        memset((void *)&quotient_large[0], 0x00, scalar_bytes * 4 * n);
+                    }
+
+                    void reset() {
+                        wire_ffts.clear();
+
+                        opening_poly = math::polynomial(n, n);
+
+                        math::polynomial w_1_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+                        math::polynomial w_2_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+                        math::polynomial w_3_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+                        math::polynomial w_4_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+                        math::polynomial z_fft = math::polynomial(4 * n + 4, 4 * n + 4);
+
+                        memset((void *)&w_1_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&w_2_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&w_3_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&w_4_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
+                        memset((void *)&z_fft[0], 0x00, sizeof(scalar_bytes) * (4 * n + 4));
```

Comment:
Can we really do this in here?

Reply:
Kate PLONK system seems to be not required at the moment. We decided to leave it in this state for some time.